File: Arrays/AddOneToNumber.cpp
/*

https://www.interviewbit.com/problems/add-one-to-number/



Given a non-negative number represented as an array of digits,

add 1 to the number ( increment the number represented by the digits ).

The digits are stored such that the most significant digit is at the head of the list.

Example:

If the vector has [1, 2, 3]

the returned vector should be [1, 2, 4]

as 123 + 1 = 124.

*/


// # 1st method (May contains leading 0's)


vector<int> Solution::plusOne(vector<int> &A) 
{
    // Do not write main() function.
    // Do not read input, instead use the arguments to the function.
    // Do not print the output, instead return values as specified
    // Still have a doubt. Checkout www.interviewbit.com/pages/sample_codes/ for more details

    int n = A.size();
    for(int i=n-1; i>=0; i--)
    {
        if(A[i] < 9) 
        {
            A[i]++;
            return A;
        }

        A[i] = 0;
    }

    vector<int> v(n+1);
    v[0] = 1;

    return v; // O(n), O(1)
    
}



// # 2nd method


vector<int> Solution::plusOne(vector<int> &A) 
{
    // Do not write main() function.
    // Do not read input, instead use the arguments to the function.
    // Do not print the output, instead return values as specified
    // Still have a doubt. Checkout www.interviewbit.com/pages/sample_codes/ for more details

    int carry=1;
    vector<int> v;
    
    for(int i=A.size()-1;i>=0;--i)
    {
        int sum;
        sum=A[i]+carry;
        carry=sum/10;
        v.push_back(sum%10);
    }
    
    v.push_back(carry);
    int i=v.size()-1;
    
    vector<int> res;
    while(i>=0 && v[i]==0)
    {
        i--;
    }
    
    while(i>=0)
    {
        res.push_back(v[i]);
        i--;
    }
    
    return res; // O(n), O(n)
}

File: Arrays/AddOneToNumber.py
""" 
https://www.interviewbit.com/problems/add-one-to-number/


Given a non-negative number represented as an array of digits, add 1 to the number ( increment the number represented by the digits ).

The digits are stored such that the most significant digit is at the head of the list.

NOTE: Certain things are intentionally left unclear in this question which you should practice asking the interviewer. For example: 
for this problem, following are some good questions to ask :

Q : Can the input have 0's before the most significant digit. Or in other words, is 0 1 2 3 a valid input?
A : For the purpose of this question, YES
Q : Can the output have 0's before the most significant digit? Or in other words, is 0 1 2 4 a valid output?
A : For the purpose of this question, NO. Even if the input has zeroes before the most significant digit.


Problem Constraints
1 <= |A| <= 106
0 <= Ai <= 9


Input Format
First argument is an array of digits.



Output Format
Return the array of digits after adding one.



Example Input
Input 1:

[1, 2, 3]


Example Output
Output 1:

[1, 2, 4]


Example Explanation
Explanation 1:

Given vector is [1, 2, 3].
The returned vector should be [1, 2, 4] as 123 + 1 = 124.
"""


class Solution:
    # @param A : list of integers
    # @return a list of integers
    def plusOne(self, A):
        # edge case: remove leading zeros
        i = 0
        while i < len(A) - 1 and A[i] == 0:
            i += 1
        A = A[i:]
        
        n = len(A)
        for i in range(n-1, -1, -1):
            if A[i] < 9:
                A[i] += 1
                return A
            A[i] = 0
        
        return [1] + A
